client/go/outline/tun2socks/tunnel.go

```diff
@@ -74,31 +73,46 @@ func newTunnel(streamDialer transport.StreamDialer, packetListener transport.Pac
 	})
 	lwipStack := core.NewLWIPStack()
 	base := tunnel.NewTunnel(tunWriter, lwipStack)
-	t := &outlinetunnel{base, lwipStack, streamDialer, packetListener, isUDPEnabled}
-	t.registerConnectionHandlers()
+	udpHandler := &toggleUDPConnHandler{
+		Handler:         NewUDPHandler(packetListener, 30*time.Second),
+		FallbackHandler: dnsfallback.NewUDPHandler(),
+	}
+	udpHandler.IsEnabled.Store(isUDPEnabled)
+	t := &outlinetunnel{base, packetListener, udpHandler}
+	core.RegisterTCPConnHandler(NewTCPHandler(streamDialer))
+	core.RegisterUDPConnHandler(udpHandler)
 	return t, nil
 }
 
 func (t *outlinetunnel) UpdateUDPSupport() bool {
 	resolverAddr := &net.UDPAddr{IP: net.ParseIP("1.1.1.1"), Port: 53}
 	isUDPEnabled := connectivity.CheckUDPConnectivityWithDNS(t.packetDialer, resolverAddr) == nil
-	if t.isUDPEnabled != isUDPEnabled {
-		t.isUDPEnabled = isUDPEnabled
-		t.lwipStack.Close() // Close existing connections to avoid using the previous handlers.
-		t.registerConnectionHandlers()
-	}
+	t.udpHandler.IsEnabled.Store(isUDPEnabled)
 	return isUDPEnabled
 }
 
-// Registers UDP and TCP connection handlers to the tunnel's host and port.
-// Registers a DNS/TCP fallback UDP handler when UDP is disabled.
-func (t *outlinetunnel) registerConnectionHandlers() {
-	var udpHandler core.UDPConnHandler
-	if t.isUDPEnabled {
-		udpHandler = NewUDPHandler(t.packetDialer, 30*time.Second)
+type toggleUDPConnHandler struct {
+	IsEnabled       atomic.Bool
```

Comment:
Maybe reverse the condition, call it `UseFallback`? So the default value will always use `Handler` instead of `FallbackHandler`.

Reply:
Done in https://github.com/Jigsaw-Code/outline-apps/pull/2479/commits/28880a92c596d56e4ae29388b0356507aaf3b670